sql: include current user in EXPLAIN bundle

spilchen opened the pull request:
Include the current user in the env.sql file of the EXPLAIN bundle. This information is useful for debugging issues related to row-level security (RLS), as the user determines which policies are applied to a query.

This change also verifies the presence of other relevant RLS metadata in the bundle, including:
- the list of existing policies,
- whether RLS is enabled on the target table,
- the policies enforced in the query plan (with EXPLAIN (VERBOSE)), and
- modifications to the plan such as injected filters that enforce RLS policies.

Includes a new test to validate these additions.

Informs #139329

Epic: CRDB-11724
Release note: none

yuzefovich reviewed (commented):
Nice! I have a couple of comments / questions.

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143671#-:-OMV9Kkl52nCl3N1u23y:bgj7l9y)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/explain_bundle.go` line 556 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143671#-OMV8Cbr4VuR16Qp8SME:-OMV8Cbr4VuR16Qp8SMF:b-oozo6a) ([raw file](https://github.com/cockroachdb/cockroach/blob/204d78e9e917ff0f9f357e0f60323a4201edb228/pkg/sql/explain_bundle.go#L556)):*
> ```Go
> 	fmt.Fprintf(&buf, "\n")
> 
> 	c.PrintUser(&buf)
> ```

Do you think we should include the stmt(s) to actually create the user / role into the bundle too? We try to make our bundles easily recreateable (via `cockroach debug sb recreate` command), and I think right now the bundle recreation will fail on a "custom" user. Perhaps we actually want to just automatically comment out `SET role ...` stmt in `env.sql`?

BTW we have a handful of `test_sb_recreate*` acceptance docker cli tests for this.
___
*[`pkg/sql/explain_bundle.go` line 1017 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143671#-OMV7h2v0XQtGFh1KmR7:-OMV7h2v0XQtGFh1KmR8:b-bxsc4d) ([raw file](https://github.com/cockroachdb/cockroach/blob/204d78e9e917ff0f9f357e0f60323a4201edb228/pkg/sql/explain_bundle.go#L1017)):*
> ```Go
> 
> func (c *stmtEnvCollector) PrintUser(w io.Writer) {
> 	fmt.Fprintf(w, "-- User: %s\n", c.p.User())
> ```

nit: double-checking - is "user" is different from current "role" ? We will already include `role` session variable in the form like
```sql
SET role = rls_user;  -- default value: none
```
into `env.sql` (we include all session variables that differ from their defaults, in addition to a handful of read-only session vars).
___
*[`pkg/sql/explain_bundle_test.go` line 986 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143671#-OMV9DUy04Bw96WeCnFN:-OMV9DUy04Bw96WeCnFO:b-8tyin1) ([raw file](https://github.com/cockroachdb/cockroach/blob/204d78e9e917ff0f9f357e0f60323a4201edb228/pkg/sql/explain_bundle_test.go#L986)):*
> ```Go
> 		)
> 
> 		r.Exec(t, "SET ROLE root")
> ```

nit: might be a bit cleaner to `defer` this after setting the role to `rls_user` above.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143671)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/explain_bundle.go` line 556 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143671#-OMV8Cbr4VuR16Qp8SME:-OMgZtvhCb4h-W40Iysa:bipahgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/204d78e9e917ff0f9f357e0f60323a4201edb228/pkg/sql/explain_bundle.go#L556)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Do you think we should include the stmt(s) to actually create the user / role into the bundle too? We try to make our bundles easily recreateable (via `cockroach debug sb recreate` command), and I think right now the bundle recreation will fail on a "custom" user. Perhaps we actually want to just automatically comment out `SET role ...` stmt in `env.sql`?

BTW we have a handful of `test_sb_recreate*` acceptance docker cli tests for this.
</blockquote></details>

Good point — I decided to comment out the `SET ROLE` statement. While we could probably recreate the user easily, the key part for RLS is the role membership, which can be tricky to reconstruct due to its hierarchical nature. So I'm not sure recreating the user adds much value.

Thanks for the tip on the test! I confirmed that `recreate` runs into issues when a custom role is set, so I added a test to cover that case.
___
*[`pkg/sql/explain_bundle.go` line 1017 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143671#-OMV7h2v0XQtGFh1KmR7:-OMga0xT-8N5aLU0JVeF:b-rnhdli) ([raw file](https://github.com/cockroachdb/cockroach/blob/204d78e9e917ff0f9f357e0f60323a4201edb228/pkg/sql/explain_bundle.go#L1017)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: double-checking - is "user" is different from current "role" ? We will already include `role` session variable in the form like
```sql
SET role = rls_user;  -- default value: none
```
into `env.sql` (we include all session variables that differ from their defaults, in addition to a handful of read-only session vars).
</blockquote></details>

Yeah, the connected user is distinct from the current role. I double-checked that if you connect as user `x` without any `SET ROLE` commands, there's no mention of the user in `env.sql` prior to my change.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: include current user in EXPLAIN bundle

Include the current user in the env.sql file of the EXPLAIN bundle.
This information is useful for debugging issues related to row-level
security (RLS), as the user determines which policies are applied to a
query.

This change also verifies the presence of other relevant RLS metadata in
the bundle, including:
- the list of existing policies,
- whether RLS is enabled on the target table,
- the policies enforced in the query plan (with EXPLAIN (VERBOSE)), and
- modifications to the plan such as injected filters that enforce RLS
policies.

Includes a new test to validate these additions.

Informs #139329

Epic: CRDB-11724
Release note: none

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 4 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/143671#-:-OMh8RuH0REJ3PrQ6XVo:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/explain_bundle.go` line 556 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/143671#-OMV8Cbr4VuR16Qp8SME:-OMh80nkESvfADM5v6yg:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/204d78e9e917ff0f9f357e0f60323a4201edb228/pkg/sql/explain_bundle.go#L556)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Good point — I decided to comment out the `SET ROLE` statement. While we could probably recreate the user easily, the key part for RLS is the role membership, which can be tricky to reconstruct due to its hierarchical nature. So I'm not sure recreating the user adds much value.

Thanks for the tip on the test! I confirmed that `recreate` runs into issues when a custom role is set, so I added a test to cover that case.
</blockquote></details>

Thanks!


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR!

bors r+
